docs: restructure working PR (#4866)
dfinity/motoko: doc/md/fundamentals/3-declarations/2-function-declarations.md

---
sidebar_position: 2
---

# Functions declarations

> A function in Motoko is a reusable block of code that takes inputs, processes them, then returns a result. Functions can be named or anonymous, and can optionally specify parameter and return types.

Reviewer: ```suggestion
A function in Motoko is a block of code that takes inputs, processes them, then (optionally) returns a result. Functions can be named or anonymous, and can optionally specify parameter and return types.
```

## Declaring a function

> Functions are declared using the `func` keyword. A named function assigns a function to an identifier, allowing recursion and reuse.

Reviewer: ```suggestion
Functions are declared using the `func` keyword. A named function assigns a function to an identifier, allowing recursion and reuse. Anonymous functions do not have a name (e.g., lambda functions for function expressions). They can be defined inline and usually assigned to variables or passed as arguments.
```

```motoko no-repl
// The function is named 'add'
// 'a: Int' and 'b: Int' are parameters with types.
// 'Int' is the return type.
func add(a: Int, b: Int) : Int {
    // Exits the function and provides a result.
    return a + b;
}
```

## Calling a function

To execute a function, pass the required arguments:

```motoko no-repl
func add(a: Int, b: Int) : Int {
    return a + b;
};
add(3, 5);
```

> ## Functions as values

Reviewer: ```suggestion
## Functions as values (anonymous functions)
```

> Functions in Motoko are first-class values. They can be assigned to variables, stored in data structures, and passed as arguments.

Reviewer: ```suggestion
Functions in Motoko are first-class values. They can be assigned to variables, stored in data structures, and passed as parameters.
```

```motoko no-repl
let double = func (x: Int) : Int { x * 2 };
let applyFunction = func (f: (Int) -> Int, value: Int) : Int { f(value) };

applyFunction(double, 10);
```

> `applyFunction` takes a function `f` and applies it to `value`.  `double` is passed as an argument, demonstrating higher-order functions.

Reviewer: ```suggestion
`applyFunction` takes a function `f` and applies it to `value`.  `double` is passed as an parameter, allowing to construct higher-order functions.
```

## Anonymous functions

Motoko supports functions without names (anonymous). They can be used in expressions or passed as arguments.

> ```motoko no-repl

Reviewer 2: A better example might be func twice(f) : Int -> Int = func (x) { f(f(x)) };

This take a function as an argument and returns an anonymous function as its result.
twice(func x = x+1)(2) returns 4. or something.

Author: @Demali-876

Reviewer 3: Resolved in #5062

Author: Moved to respective PR

let multiply = func (x: Int, y: Int) : Int { x * y };
multiply(4, 2);
```

The function is assigned to `multiply` instead of being named directly. `{ x * y }` is the function body without `return`, as single expressions are implicitly returned.

## Recursive functions

> A function that can call itself is a recursive function. Recursion allows looping behavior without explicit loops.

Reviewer: ```suggestion
A function that can call itself is a recursive function. Recursion allows looping behavior without explicit loops. NB: recursion implies allocating stack frames, which allocates additional memory on the (relatively small) program stack. Too deep recursion may result in out of memory errors. 
```

```motoko no-repl
func factorial(n: Nat) : Nat {
    if (n == 0) {
        return 1;
    };
// Calls itself with factorial(n - 1), reducing n until reaching 0.
    return n * factorial(n - 1);
}
```

> 

Reviewer 2: I think we need a section on generic functions:

maybe just Twice<T>(f: T -> T) : T ->T {  f(f(x) } would do

Author: @Demali-876

Reviewer 3: Resolved in #5062

## Shared functions in actors

> In actors, functions can be marked as `shared` to allow [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) [inter-canister](https://internetcomputer.org/docs/motoko/fundamentals/messaging) communication.

Author: The docs are not live yet, therefore the links do not work yet.

> In actors, functions can be marked as `shared` to allow [asynchronous](https://internetcomputer.org/docs/motoko/fundamentals/actors-async#async--await) [inter-canister](https://internetcomputer.org/docs/motoko/fundamentals/messaging) communication.

Reviewer 2: Is there any reason these links aren't just  markdown links to local files? Then they would more likely work with the `make preview`?

```motoko no-repl
actor Counter {
    stable var count: Nat = 0;

    public func inc() : async () {
        count += 1;
    };

    public shared query func getCount() : async Nat {
        return count;
    }
};
await Counter.inc();
await Counter.getCount();
```

### Accessing the caller identity

> One key advantage of `shared` functions is that they can access the caller's `Principal`, which represents the entity (user or canister) making the request. This allows functions to enforce access control or record who invoked them.

Reviewer: ```suggestion
One key advantage of `shared` functions is that they can access the caller's `Principal`, which represents the entity (user or canister) making the request. This enables actors to enforce access control or record who invoked them.
```
